src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentManager.cs

```diff
@@ -0,0 +1,133 @@
+
+            if (documentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(documentFactory));
+            }
+
+            _joinableTaskContext = joinableTaskContext;
+            _fileUriProvider = fileUriProvider;
+            _documentFactory = documentFactory;
+            _documents = new Dictionary<Uri, DocumentTracker>();
+        }
+
+        public override void TrackDocumentView(ITextBuffer buffer, ITextView textView)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (textView is null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            Debug.Assert(_joinableTaskContext.IsOnMainThread);
+
+            var uri = _fileUriProvider.GetOrCreate(buffer);
+            if (!_documents.TryGetValue(uri, out var documentTracker))
+            {
+                var lspDocument = _documentFactory.Create(buffer);
+                documentTracker = new DocumentTracker(lspDocument);
+                _documents[uri] = documentTracker;
+            }
+
+            documentTracker.TextViews.Add(textView);
+        }
+
+        public override void UntrackDocumentView(ITextBuffer buffer, ITextView textView)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (textView is null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            Debug.Assert(_joinableTaskContext.IsOnMainThread);
+
+            var uri = _fileUriProvider.GetOrCreate(buffer);
+            if (!_documents.TryGetValue(uri, out var documentTracker))
+            {
+                // We don't know about this document, noop.
+                return;
+            }
+
+            documentTracker.TextViews.Remove(textView);
+
+            if (documentTracker.TextViews.Count == 0)
+            {
+                _documents.Remove(uri);
+            }
+        }
+
+        public override bool TryGetDocument(Uri uri, out LSPDocument lspDocument)
+        {
+            Debug.Assert(_joinableTaskContext.IsOnMainThread);
+
+            if (!_documents.TryGetValue(uri, out var documentTracker))
+            {
+                // This should never happen in practice but return `null` so our tests can validate
+                lspDocument = null;
+                return false;
+            }
+
+            lspDocument = documentTracker.Document;
+            return true;
+        }
+
+        private class DocumentTracker
+        {
+            public DocumentTracker(LSPDocument document)
+            {
+                Document = document;
+                TextViews = new HashSet<ITextView>();
+            }
+
+            public LSPDocument Document { get; }
+
+            public HashSet<ITextView> TextViews { get; }
```

Comment:
>HashSet [](start = 19, length = 7)

This seems a bit heavy for a collection that almost always has one entry.

Reply:
This'll go away when we move to ITextDocumentFactoryService lifecycle tracking